### PR TITLE
sql: use constraint name when adding a primary key constraint

### LINKED ISSUE #52833: sql: constraint name is not respected when adding a primary key constraint
1. Create a table without an explicit primary key
2. Add a primary key with a named constraint

Expected: The primary key constraint should have the specified name.
Actual: The primary key constraint has the name `"primary"`.

```
root@:26257/defaultdb> create table a (a int not null);
CREATE TABLE

root@:26257/defaultdb> alter table a add constraint my_pk primary key (a);
NOTICE: primary key changes are finalized asynchronously; further schema changes on this table may be restricted until the job completes
ALTER TABLE

root@:26257/defaultdb> show create table a;
  table_name |               create_statement
-------------+------------------------------------------------
  a          | CREATE TABLE public.a (
             |     a INT8 NOT NULL,
             |     CONSTRAINT "primary" PRIMARY KEY (a ASC),
             |     FAMILY "primary" (a, rowid)
             | )
(1 row)
```

### PR DESCRIPTION
**Pull request by neeral**
fixes https://github.com/cockroachdb/cockroach/issues/52833

Previously, when a table is created without a primary key and one
is later added by a named constraint, the specified name was
ignored. The primary key constraint always had the name "primary".

This was inadequate because the specified name was ignored. It
causes some trouble for schema migration tools.

To address this, this patch names the primary key as specified in
the `ALTER TABLE <t> ADD CONSTAINT <name> PRIMARY KEY` statement.

Release (sql change, bug fix): Adding a primary key constraint
to a table without a primary key no longer ignores the name
specified for the primary key.

### CODE REVIEW

**Review by neeral — COMMENTED** [review]
@rafiss @solongordon this PR is ready for you to review, its passed TC builds. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57146#-:-MN4RQdu9ADuKkerEiHf:b-qo3k3c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @solongordon)

File: pkg/sql/alter_primary_key.go [hunk 1]
```diff
@@ -129,6 +129,12 @@ func (p *planner) AlterPrimaryKey(
 		"new_primary_key",
 		nameExists,
 	)
+	if alterPKNode.Name != "" &&
+		nameExists(string(alterPKNode.Name)) &&
+		// Allow reuse of existing primary key's name.
+		tableDesc.PrimaryIndex.Name != string(alterPKNode.Name) {
```
Comment (rafiss) [inline pkg/sql/alter_primary_key.go:135]: if you swap the orders of the `nameExists` check and the equality check, it should be slightly better for short-circuiting

File: pkg/sql/logictest/testdata/logic_test/alter_primary_key [hunk 2]
```diff
@@ -726,14 +726,14 @@ t  CREATE TABLE public.t (
 statement ok
 DROP TABLE IF EXISTS t;
 CREATE TABLE t (x INT NOT NULL);
-ALTER TABLE t ADD CONSTRAINT "primary" PRIMARY KEY (x)
+ALTER TABLE t ADD CONSTRAINT "my_pk" PRIMARY KEY (x)
```
Comment (rafiss) [inline pkg/sql/logictest/testdata/logic_test/alter_primary_key:730]: can you also add a test to make sure "name already exists" error is returned correctly?

File: pkg/sql/catalog/tabledesc/structured.go [hunk 3]
```diff
@@ -3236,7 +3236,11 @@ func (desc *Mutable) MakeMutationComplete(m descpb.DescriptorMutation) error {
 			if err != nil {
 				return err
 			}
-			newIndex.Name = "primary"
+			if args.NewPrimaryIndexName == "" {
+				newIndex.Name = "primary"
```
Comment (rafiss) [inline pkg/sql/catalog/tabledesc/structured.go:3240]: can you change this to use the `PrimaryKeyIndexName` constant?

File: pkg/sql/alter_primary_key.go [hunk 4]
```diff
@@ -129,6 +129,12 @@ func (p *planner) AlterPrimaryKey(
...
+		return errors.Newf("constraint with name %s already exists", alterPKNode.Name)
```
Comment (rafiss) [inline pkg/sql/alter_primary_key.go:136]: is there an appropriate PG error code for this? see `pgcode/codes.go` and `pgerror.Newf()`

**Review by rafiss — CHANGES_REQUESTED** [review]
thank you! overall this looks good, but i have a few comments to address

File: pkg/sql/alter_primary_key.go (hunk 4 above)
Comment (neeral) [reply pkg/sql/alter_primary_key.go]: Thanks for the tip. I used the error code for InvalidName.

**Review by neeral — COMMENTED** [review]
Thanks for the review @rafiss, I've addressed the four comments that you made.
PTAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57146#-:-MNjg22W2UaJVG-DeZXV:bonq5v3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @neeral, @rafiss, and @solongordon)

**Comment by neeral** [comment]
@rafiss could you PTAL ?

**Review by rafiss — CHANGES_REQUESTED** [review]
@neeral sorry for the delay! this looks very good. i just have a small comment, then this can be merged

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57146#-:-MOCl23h8V1uDJ-Qtd6X:byc6y2w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @neeral and @solongordon)

___
*[pkg/sql/alter_primary_key.go, line 136 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57146#-MN0IgqBwK0D6Xa5Kp6z-r1-136:-MOCko-o9FsiEU5G8Rxo:b-wdn7dp) ([raw file](https://github.com/cockroachdb/cockroach/blob/81c17cc484b6ba4983e835f88a239910433d0281/pkg/sql/alter_primary_key.go#L136)):*
<details><summary><i>Previously, neeral (Neeral Dodhia) wrote…</i></summary><blockquote>

Thanks for the tip. I used the error code for InvalidName.
</blockquote></details>

sorry i was just able to read through the list and figure out how we use different codes. I think `42710` (duplicate_object) is a better error code here.

**Review by neeral — COMMENTED** [review]
TFTR @rafiss , much appreciated. I've updated the pgerror code as you suggested.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57146#-:-MODBwpH3JLr25T0D4ID:bhnomth)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @neeral, @rafiss, and @solongordon)

___
*[pkg/sql/alter_primary_key.go, line 136 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57146#-MN0IgqBwK0D6Xa5Kp6z-r1-136:-MODBtf91WSedketpdU2:b-32v6xi) ([raw file](https://github.com/cockroachdb/cockroach/blob/81c17cc484b6ba4983e835f88a239910433d0281/pkg/sql/alter_primary_key.go#L136)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

sorry i was just able to read through the list and figure out how we use different codes. I think `42710` (duplicate_object) is a better error code here.
</blockquote></details>

Thanks, updated it to duplicate_object

**Review by rafiss — APPROVED** [review]
lgtm! i will take care of merging this soon

thanks very much for your contribution :) 

> Reviewed 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57146#-:-MODZEw55nJYgpvTi6b-:b95uj85)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @solongordon)

**Comment by rafiss** [comment]
an internal server the tests rely on seems to be flaky... trying again

bors r+

**Comment by rafiss** [comment]
the internal server is fixed now

bors r+

**Comment by neeral** [comment]
Thanks for pushing this through @rafiss , especially with the build flakiness.

**Pull request merged**